### PR TITLE
Update documentation on authoring type providers

### PR DESCRIPTION
**Pull request by dsyme**
# Revamp documentation for authoring type providers

## Summary

The document on authoring type providers was an out-of-date mess with incorrect formatting.  This gets it closer to par

## Details

* Update code to match latest TypeProvider SDK
* Clean up indentation of code (which was nearly all wildly off)
* List three different type providers

## Suggested Reviewers

@cartermp review please

### CODE REVIEW

**Comment by dsyme** [comment]
The document was a sufficient mess that we should probably just merge fairly quickly then refine (so we can see it in situ)

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: nit: we're using https links whenever they're accepted. I just fixed all of them in the F# directory yesterday.

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: extra https in the first link.

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: you'll need to add two spaces after this link so the new link shows up in a new line

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: no punctuation in headings

File: docs/fsharp/tutorials/type-providers/creating-a-type-provider.md
Comment (mairaw) [inline docs/fsharp/tutorials/type-providers/creating-a-type-provider.md]: ideally the apis would use xref links to go to the API docs:
`<xref:System.NotSupportedException>`, etc. 

Remove extra space after the last comma

**Review by mairaw — COMMENTED** [review]
Left a few comments.

Re. this:
> The document was a sufficient mess that we should probably just merge fairly quickly then refine (so we can see it in situ)

You'll be able to see how the content looks like on the staging site before merging. My preference would be for us to validate that staging looks good before merging.

**Comment by dsyme** [comment]
@mairaw Super, thanks!

> You'll be able to see how the content looks like on the staging site before merging. My preference would be for us to validate that staging looks good before merging.

OK, great.

**Review by cartermp — APPROVED** [review]
LGTM, pending it "looking right" on staging. @mairaw we used to have a bot which would give us the link here. Can we access it through some other means?

**Comment by mairaw** [comment]
@cartermp the build checks here on the PR include a link to the build report. Click on Details from the OpenPublishing.Build check.

Sometimes, the build will have and only have the topics you changed in there and sometimes like right now, it shows a bunch of crap. So you have to open one of those links and paste the path you want to validate. 😢 

Here's the direct link: https://review.docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider?branch=pr-en-us-4518

Feel free to merge it if you think it's ready @cartermp.

**Comment by dsyme** [comment]
I don't seem to have access to https://review.docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider?branch=pr-en-us-4518.  I tried signing in with my Microsoft account.  Should I use a different account?

**Comment by dsyme** [comment]
OK, I can access the review site, it LGTM

**Comment by cartermp** [comment]
Ditto. Thanks @dsyme!

**Pull request merged**
